### PR TITLE
Python: improve site_packages_dir handling

### LINKED ISSUE #15304: ParaView: configuration can't find mpi4py with system python
### Spack version

```console
$ spack --version
0.13.2
```

### Steps to reproduce the issue

```console
$ spack install paraview@5.6.2%intel@19.0.4 +osmesa+opengl2+python3+plugins ^mesa%gcc@7.4.0
     635    -- Check size of void* - done
     636    Traceback (most recent call last):
     637      File "<string>", line 1, in <module>
     638    ModuleNotFoundError: No module named 'mpi4py'
  >> 639    CMake Error at VTK/Parallel/MPI4Py/CMakeLists.txt:23 (include_directories):
     640      include_directories given empty-string as include directory.
```

### Platform and user environment

```commandline
$ uname -a 
Linux sn-fey2.lanl.gov 3.10.0-1062.1.1.1chaos.ch6.x86_64 #1 SMP Wed Sep 4 16:09:20 PDT 2019 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -d
Description:	Red Hat Enterprise Linux Server release 7.7 (Maipo)
``` 

- `packages.yaml`
```yaml
  python:
    paths:
      python@3.6.8: /usr
    buildable: False
```
- `compilers.yaml`
```yaml
- compiler:
    environment: {}
    extra_rpaths:
      - /usr/projects/hpcsoft/toss3/common/intel-clusterstudio/2019.4.070/compilers_and_libraries_2019/linux/lib/intel64
    flags: {}
    modules: [friendly-testing, intel/19.0.4]
    operating_system: rhel7
    paths:
      cc: icc
      cxx: icpc
      f77: ifort
      fc: ifort
    spec: intel@19.0.4
    target: x86_64
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: [gcc/7.4.0]
    operating_system: rhel6
    paths:
      cc: gcc
      cxx: g++
      f77: gfortran
      fc: gfortran
    spec: gcc@7.4.0
    target: x86_64
```

### Additional information

@danlipsa @chuckatkins

### LINKED ISSUE #20043: Installation issue: py-matplolib conflicts with system matplotlib
The following error seems to exist if there is a version of matplotlib installed with the system package manager. The spack version of the package does not work as expected. Based on the information [here](https://stackoverflow.com/questions/61864151/keyerror-scale-for-import-matplotlib-pyplot-as-plt), possibly spack isn't correctly linking the mpl_toolkits folder?

### Steps to reproduce the issue

```console
$ spack install py-matplotlib
$ spack load py-matplotlib
$ python3 -c "import matplotlib.pyplot"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/mersoj/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/py-matplotlib-3.3.3-hoj5zbtgg6ayuek2ce6sjouhqhh2pkkq/lib/python3.8/site-packages/matplotlib/pyplot.py", line 43, in <module>
    from matplotlib.figure import Figure, figaspect
  File "/home/mersoj/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/py-matplotlib-3.3.3-hoj5zbtgg6ayuek2ce6sjouhqhh2pkkq/lib/python3.8/site-packages/matplotlib/figure.py", line 18, in <module>
    from mpl_toolkits.mplot3d import Axes3D
  File "/usr/lib/python3/dist-packages/mpl_toolkits/mplot3d/__init__.py", line 1, in <module>
    from .axes3d import Axes3D
  File "/usr/lib/python3/dist-packages/mpl_toolkits/mplot3d/axes3d.py", line 42, in <module>
    class Axes3D(Axes):
  File "/usr/lib/python3/dist-packages/mpl_toolkits/mplot3d/axes3d.py", line 50, in Axes3D
    def __init__(
  File "/home/mersoj/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/py-matplotlib-3.3.3-hoj5zbtgg6ayuek2ce6sjouhqhh2pkkq/lib/python3.8/site-packages/matplotlib/docstring.py", line 80, in dedent_interpd
    return interpd(func)
  File "/home/mersoj/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/py-matplotlib-3.3.3-hoj5zbtgg6ayuek2ce6sjouhqhh2pkkq/lib/python3.8/site-packages/matplotlib/docstring.py", line 40, in __call__
    func.__doc__ %= self.params
KeyError: 'scale'
```

### Information on your system

* **Spack:** 0.16.0-31-a67bf0ca7
* **Python:** 3.8.5
* **Platform:** linux-ubuntu20.04-broadwell
* **Concretizer:** original

python is external system python 3.8.5 

@adamjstewart

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [N/A] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

### LINKED ISSUE #22299: ModuleNotFoundError: No module named 'numpy' when py-numcodecs@0.7.3 pytest
Hello,I meet a problem about ModuleNotFoundError: No module named 'numpy' when py-numcodecs@0.7.3 pytest.
### Steps to reproduce the issue

```console
localhost:/tmp/root/spack-stage/spack-stage-py-numcodecs-0.7.3-mdm2ofae7zyu6ymyvqmhpiixwjcfjqpa/spack-src #spack load py-numcodecs@0.7.3
localhost:/tmp/root/spack-stage/spack-stage-py-numcodecs-0.7.3-mdm2ofae7zyu6ymyvqmhpiixwjcfjqpa/spack-src #spack load py-pip
localhost:/tmp/root/spack-stage/spack-stage-py-numcodecs-0.7.3-mdm2ofae7zyu6ymyvqmhpiixwjcfjqpa/spack-src # ls
CODE_OF_CONDUCT.md  MANIFEST.in  README.rst   adhoc  build.cmd  cpuinfo.py  fixture    numcodecs           pytest.ini        requirements_dev.txt   requirements_test.txt  setup.py
LICENSE             PKG-INFO     __pycache__  build  c-blosc    docs        notebooks  numcodecs.egg-info  requirements.txt  requirements_rtfd.txt  setup.cfg              tox.ini
localhost:/tmp/root/spack-stage/spack-stage-py-numcodecs-0.7.3-mdm2ofae7zyu6ymyvqmhpiixwjcfjqpa/spack-src #pip install -r requirements.txt
localhost:/tmp/root/spack-stage/spack-stage-py-numcodecs-0.7.3-mdm2ofae7zyu6ymyvqmhpiixwjcfjqpa/spack-src #pip install -r requirements_test.txt
localhost:/tmp/root/spack-stage/spack-stage-py-numcodecs-0.7.3-mdm2ofae7zyu6ymyvqmhpiixwjcfjqpa/spack-src #python setup.py test
localhost:/tmp/root/spack-stage/spack-stage-py-numcodecs-0.7.3-mdm2ofae7zyu6ymyvqmhpiixwjcfjqpa/spack-src #pytest           
……
    from numcodecs.zlib import Zlib
numcodecs/zlib.py:5: in <module>
    from .compat import ndarray_copy, ensure_contiguous_ndarray
numcodecs/compat.py:7: in <module>
    import numpy as np
E   ModuleNotFoundError: No module named 'numpy'
_________________________________________________________________________________ ERROR collecting numcodecs/tests/test_zlib.py _________________________________________________________________________________
ImportError while importing test module '/tmp/root/spack-stage/spack-stage-py-numcodecs-0.7.3-mdm2ofae7zyu6ymyvqmhpiixwjcfjqpa/spack-src/numcodecs/tests/test_zlib.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/home/spack/opt/spack/linux-sles15-aarch64/gcc-7.4.0/python-3.7.7-tmyoucv3qh4ffcqbrqc7wcehar7a2n3w/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
numcodecs/__init__.py:27: in <module>
    from numcodecs.zlib import Zlib
numcodecs/zlib.py:5: in <module>
    from .compat import ndarray_copy, ensure_contiguous_ndarray
numcodecs/compat.py:7: in <module>
    import numpy as np
E   ModuleNotFoundError: No module named 'numpy'
_________________________________________________________________________________ ERROR collecting numcodecs/tests/test_zstd.py _________________________________________________________________________________
ImportError while importing test module '/tmp/root/spack-stage/spack-stage-py-numcodecs-0.7.3-mdm2ofae7zyu6ymyvqmhpiixwjcfjqpa/spack-src/numcodecs/tests/test_zstd.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/home/spack/opt/spack/linux-sles15-aarch64/gcc-7.4.0/python-3.7.7-tmyoucv3qh4ffcqbrqc7wcehar7a2n3w/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
numcodecs/__init__.py:27: in <module>
    from numcodecs.zlib import Zlib
numcodecs/zlib.py:5: in <module>
    from .compat import ndarray_copy, ensure_contiguous_ndarray
numcodecs/compat.py:7: in <module>
    import numpy as np
E   ModuleNotFoundError: No module named 'numpy'
============================================================================================ short test summary info ============================================================================================
ERROR numcodecs/tests/test_astype.py
ERROR numcodecs/tests/test_base64.py
ERROR numcodecs/tests/test_blosc.py
ERROR numcodecs/tests/test_bz2.py
ERROR numcodecs/tests/test_categorize.py
ERROR numcodecs/tests/test_checksum32.py
ERROR numcodecs/tests/test_compat.py
ERROR numcodecs/tests/test_delta.py
ERROR numcodecs/tests/test_fixedscaleoffset.py
ERROR numcodecs/tests/test_gzip.py
ERROR numcodecs/tests/test_json.py
ERROR numcodecs/tests/test_lz4.py
ERROR numcodecs/tests/test_lzma.py
ERROR numcodecs/tests/test_msgpacks.py
ERROR numcodecs/tests/test_packbits.py
ERROR numcodecs/tests/test_pickles.py
ERROR numcodecs/tests/test_quantize.py
ERROR numcodecs/tests/test_registry.py
ERROR numcodecs/tests/test_vlen_array.py
ERROR numcodecs/tests/test_vlen_bytes.py
ERROR numcodecs/tests/test_vlen_utf8.py
ERROR numcodecs/tests/test_zlib.py
ERROR numcodecs/tests/test_zstd.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 23 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

localhost:/tmp/root/spack-stage/spack-stage-py-numcodecs-0.7.3-mdm2ofae7zyu6ymyvqmhpiixwjcfjqpa/spack-src # spack edit py-numcodecs
# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
# Spack Project Developers. See the top-level COPYRIGHT file for details.
#
# SPDX-License-Identifier: (Apache-2.0 OR MIT)

from spack import *


class PyNumcodecs(PythonPackage):
    """Numcodecs is a Python package providing buffer compression and
    transformation codecs for use in data storage and communication
    applications.
    """

    homepage = "https://github.com/zarr-developers/numcodecs"
    url      = "https://pypi.io/packages/source/n/numcodecs/numcodecs-0.6.4.tar.gz"

    version('0.7.3', sha256='022b12ad83eb623ec53f154859d49f6ec43b15c36052fa864eaf2d9ee786dd85')
    version('0.6.4', sha256='ef4843d5db4d074e607e9b85156835c10d006afc10e175bda62ff5412fca6e4d')

    variant('msgpack', default=False, description='Codec to encode data as msgpacked bytes.')

    depends_on('py-setuptools@18.0:',       type='build')
    depends_on('py-setuptools-scm@1.5.4:',  type=('build', 'run'))
    depends_on('py-numpy@1.7:',             type=('build', 'run'))
    depends_on('py-msgpack', type=('build', 'run'), when='+msgpack')

```
I want to know why **pytest** still reports ModuleNotFoundError: No module named'numpy' when the package package contains the **numpy** dependency. Whether the path of **py-numpy** is not found during the spack setting. As a result, the **spack load py-numcodecs** does not load dependent **py-numpy**, and I need to load it again. Can you help me analyze it?
### Information on your system
localhost:/tmp/root/spack-stage/spack-stage-py-numcodecs-0.7.3-mdm2ofae7zyu6ymyvqmhpiixwjcfjqpa/spack-src # spack debug report
* **Spack:** 0.16.0-77-98460b6f47
* **Python:** 3.8.6
* **Platform:** linux-sles15-aarch64
* **Concretizer:** original

### Additional information

- [ ] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [ ] I have searched the issues of this repo and believe this is not a duplicate
- [ ] I have run the failing commands in debug mode and reported the output

### LINKED ISSUE #24076: Installation issue:  py-numpy
```console
$ spack install py-numpy
...
==> Error: ProcessError: Command exited with status 1:
    '/usr/bin/python3.8' '-s' 'setup.py' '--no-user-cfg' 'build'
See build log for details:
  /tmp/permeakra/spack-stage/spack-stage-py-numpy-1.20.3-moeytesgplajh7ebnge5ultezzwfu3h3/spack-build-out.txt

$ spack build-env py-numpy python
Python 3.8.5 (default, Jan 27 2021, 15:41:15) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import Cython
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'Cython'
```
### Information on your system

spack debug report 
* **Spack:** 0.16.2-2966-690558f927
* **Python:** 3.8.5
* **Platform:** linux-ubuntu20.04-haswell
* **Concretizer:** original

Forced use of external system python. Spack insists on installing its own Cython. Purging external cython didn't have obvious effect.

### Additional information

there were remnants of python2 in the system. Purging them didn't do anything, but maybe  I missed smth

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

### LINKED ISSUE #26546: Installation issue: meson broken on multilib systems with lib/lib64 split (like openSUSE)
### Steps to reproduce the issue

```console
$ spack install meson
$ spack load meson
$ meson --help
meson --help
Traceback (most recent call last):
  File "/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-11.2.0/meson-0.59.2-w3xwpklyb6mnoosqb6oclexad6clr2on/bin/meson", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3123, in <module>
    @_call_aside
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3107, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3136, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 578, in _build_master
    ws.require(__requires__)
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 895, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 781, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'meson==0.59.2' distribution was not found and is required by the application
$ echo $PYTHONPATH
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-11.2.0/py-setuptools-57.4.0-p7r4vvplum63eqqx6tf7qwfitk7ppalp/lib64/python3.6/site-packages:/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-11.2.0/meson-0.59.2-w3xwpklyb6mnoosqb6oclexad6clr2on/lib64/python3.6/site-packages:/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-11.2.0/py-setuptools-57.4.0-p7r4vvplum63eqqx6tf7qwfitk7ppalp/lib64/python3.6/site-packages:/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-11.2.0/ninja-1.10.2-x4i5ucjcn43khkqqjam4ae3xqcn5ahtk/misc
$ ls /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-11.2.0/meson-0.59.2-w3xwpklyb6mnoosqb6oclexad6clr2on/lib64/python3.6/site-packages
$ ls /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-11.2.0/meson-0.59.2-w3xwpklyb6mnoosqb6oclexad6clr2on/lib/python3.6/site-packages
meson-0.59.2-py3.6.egg-info  mesonbuild
```

Meaning: the directory included in `$PYTHONPATH` does not contain the Meson distro while the `lib` dir would.

@michaelkuhn

### Information on your system

* **Spack:** 0.16.3-4672-572791006d
* **Python:** 3.6.13
* **Platform:** linux-opensuse_leap15-sandybridge
* **Concretizer:** clingo

### Additional information

Since this is not a build problem, but rather that the installed files are broken I don't have the spack-build* files.
I wonder whether this is a general problem with Python packages.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### LINKED ISSUE #27497: Installation issue: py-pytest with external (system) python
### Steps to reproduce the issue

Create an environment with:
```console
$ mkdir tmp && cd tmp
$ spack env create -d .
$ spack external find python
$ spack add py-pytest
```
and then edit it to have a single external Python. In my case the final `spack.yaml` looks like:
```yaml
# This is a Spack Environment file.
#
# It describes a set of packages to be installed, along with
# configuration settings.
spack:
  # add package specs to the `specs` list
  specs:
  - py-pytest
  view: true
  packages:
    python:
      externals:
      - spec: python@3.8.10+bz2+ctypes+dbm+lzma+nis+pyexpat+readline+sqlite3+ssl~tix~tkinter+uuid+zlib
        prefix: /usr
```
When trying to install the environment you should get errors near the end:
```console
$ spack install
[ ... ]
==> Installing py-py-1.9.0-vohcpazqyqvxez6gopbq45pzxuaundmm
==> No binary for py-py-1.9.0-vohcpazqyqvxez6gopbq45pzxuaundmm found: installing from source
==> Using cached archive: /home/culpo/PycharmProjects/spack/var/spack/cache/_source-cache/archive/9c/9ca6883ce56b4e8da7e79ac18787889fa5206c79dcc67fb065376cd2fe03f342.tar.gz
==> No patches needed for py-py
==> py-py: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 1:
    '/usr/bin/python3.8' '-s' 'setup.py' '--no-user-cfg' 'build'

1 warning found in build log:
  >> 3    WARNING: The wheel package is not available.
     4    Traceback (most recent call last):
     5      File "setup.py", line 45, in <module>
     6        main()
     7      File "setup.py", line 5, in main
     8        setup(
     9      File "/home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.3.0/py-setuptools-58.2.0-j3uiv6hk77qtaa3jo2mnek6gke5wzshg/lib/p
          ython3/dist-packages/setuptools/__init__.py", line 152, in setup

See build log for details:
  /tmp/culpo/spack-stage/spack-stage-py-py-1.9.0-vohcpazqyqvxez6gopbq45pzxuaundmm/spack-build-out.txt
[ ... ]
```

### Information on your system

* **Spack:** 0.17.0-171-2f84643c26
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/7551975/spack-build-out.txt)

@adamjstewart 

Note that the issue is just with external Python. Letting Spack build a new interpreter works correctly. Being able to use a system Python is needed to let Spack bootstrap `py-test` and remove the outdated vendored dependency.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by adamjstewart**
This PR includes the following changes. I'll do my best to provide the background and motivation behind these changes. Some of these changes were extracted from #27798, which will be rebased after this is merged.

- [x] Query `sysconfig` instead of `distutils.sysconfig`

In Python 3.10, the `distutils` module has been deprecated. In Python 3.12, it will be removed entirely. To future-proof our Python build system, we should use `sysconfig` instead (added in Python 2.7 and 3.2). Also, the Debian/Ubuntu system Python contains a bug where `distutils.sysconfig` and `sysconfig` return different purelib/platlib paths. Since build backends like setuptools use `sysconfig`, we also need to use `sysconfig` for compatibility.

- [x] Add both purelib and platlib to the `PYTHONPATH`

On the majority of systems, purelib and platlib are identical, providing the location where third-party Python libraries are installed. Historically, we've been calling this `site_packages_dir`. However, when using the system Python on RHEL/CentOS/Fedora, purelib uses `lib` while platlib uses `lib64`. Different packages will end up in different directories depending on whether they contain C/C++ code or are written in pure Python. Since Spack has no way of knowing this, we should always add both purelib and platlib to the `PYTHONPATH`.

- [x] Fix bug in which dependencies are added to `PYTHONPATH`

Previously we were recursively adding all build/run/test dependencies. However, this caused build deps of build deps to be added. The correct solution is to add all direct build/run/test deps, and all recursive run deps of those.

### External references

* https://discuss.python.org/t/understanding-site-packages-directories/12959
* https://ffy00.github.io/blog/02-python-debian-and-the-install-locations/
* https://docs.python.org/3/library/sysconfig.html#installation-paths
* https://github.com/pypa/setuptools/discussions/2813

Fixes #15304
Fixes #20043
Fixes #22299
Fixes #24076
Fixes #26546
Fixes #27497

### CODE REVIEW

File: lib/spack/spack/build_systems/sip.py
Comment (adamjstewart) [inline lib/spack/spack/build_systems/sip.py]: This is actually a behavior change (platlib -> purelib). I'm not sure which is better or if both need to be used. This only affects RHEL/CentOS/Fedora system Python and sip packages and post-installation testing, so it's not a big deal. We have almost no sip packages anyway.

File: var/spack/repos/builtin/packages/precice/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/precice/package.py:146]: Some packages are installed into purelib while others are installed into platlib. In the case of numpy, I'm guessing it's platlib since it isn't just pure Python code.

Most packages assume purelib at the moment, but that probably isn't true. I actually thought about removing `site_packages_dir` entirely and forcing packages to choose between purelib or platlib, but didn't want to rock the boat too much.

File: var/spack/repos/builtin/packages/python/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/python/package.py:765]: Not worth all the error handling. For most installations, this will be overridden anyway.

File: var/spack/repos/builtin/packages/python/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/python/package.py:1068]: These are used in almost no packages so I just removed them. Like `site_packages_dir`, there isn't a single `python_lib_dir` or `python_include_dir`, there are platform-specific and platform-independent versions of these.

File: lib/spack/spack/build_systems/python.py
Comment (becker33) [inline lib/spack/spack/build_systems/python.py:131]: (Minor, only bother if you're editing the file anyway) this would be easier to read as `python_pkg`

File: lib/spack/spack/build_systems/sip.py
Comment (becker33) [reply lib/spack/spack/build_systems/sip.py]: I think it should be both, in terms of technical correctness

File: var/spack/repos/builtin/packages/precice/package.py
Comment (becker33) [reply var/spack/repos/builtin/packages/precice/package.py:146]: I think we should make that change so packages have to specify platlib vs purelib, but I agree it can be put off for future work. I don't think it's necessary for the pip PR either, so not the same sort of rush.

Comment (becker33) [reply var/spack/repos/builtin/packages/precice/package.py:146]: At the same time we should probably switch to prefix the python variables that we poke into module scope with `python_`

**Review by becker33 — DISMISSED** [review]
Some comments, may or may not be worth addressing but I'll approve and defer to you on whether to make any changes before merging.

File: lib/spack/spack/build_systems/sip.py
Comment (adamjstewart) [reply lib/spack/spack/build_systems/sip.py]: Took a look through the source code and it looks like platlib is the correct answer:
```python
        Option('target_dir', default=sysconfig.get_path('platlib'),                      
                help="the target installation directory", metavar="DIR",                 
                tools=['build', 'install']),                     
```

File: lib/spack/spack/build_systems/python.py
Comment (adamjstewart) [reply lib/spack/spack/build_systems/python.py:131]: I think I prefer `pkg` because it's shorter. It's only a local variable, so naming isn't particularly important.

**Comment by adamjstewart** [comment]
@becker33 ready for another round of review. All usage of `site_packages_dir` has been replaced by either `python_purelib` or `python_platlib`. I rather arbitrarily decided between purelib and platlib based on whether the package had non-Python dependencies or not. As to whether these packages follow that rule or not, only CI will tell...

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Review by becker33 — DISMISSED** [review]
Assuming the pipeline comes back successful, I think the platlib vs purelib decisions look reasonable.

**Comment by adamjstewart** [comment]
Failing build systems check in GitLab CI will be fixed by #28360.

**Comment by adamjstewart** [comment]
@becker33 @scheibelp is this good to go?

**Review by becker33 — APPROVED** [review]

**Pull request merged**
